src/math/trig.c

```diff
@@ -9,568 +9,97 @@
-	2081515602,
-	2083126253,
-	2084717297,
-	2086288719,
-	2087840504,
-	2089372637,
-	2090885104,
-	2092377891,
-	2093850984,
-	2095304369,
-	2096738031,
-	2098151959,
-	2099546138,
-	2100920555,
-	2102275198,
-	2103610053,
-	2104925108,
-	2106220351,
-	2107495769,
-	2108751351,
-	2109987084,
-	2111202958,
-	2112398959,
-	2113575079,
-	2114731304,
-	2115867625,
-	2116984030,
-	2118080510,
-	2119157053,
-	2120213650,
-	2121250291,
-	2122266966,
-	2123263665,
-	2124240379,
-	2125197099,
-	2126133816,
-	2127050521,
-	2127947205,
-	2128823861,
-	2129680479,
-	2130517051,
-	2131333571,
-	2132130029,
-	2132906419,
-	2133662733,
-	2134398965,
-	2135115106,
-	2135811152,
-	2136487094,
-	2137142926,
-	2137778643,
-	2138394239,
-	2138989707,
-	2139565042,
-	2140120239,
-	2140655292,
-	2141170196,
-	2141664947,
-	2142139540,
-	2142593970,
-	2143028233,
-	2143442325,
-	2143836243,
-	2144209981,
-	2144563538,
-	2144896909,
-	2145210091,
-	2145503082,
-	2145775879,
-	2146028479,
-	2146260880,
-	2146473079,
-	2146665075,
-	2146836865,
-	2146988449,
-	2147119824,
-	2147230990,
-	2147321945,
-	2147392689,
-	2147443221,
-	2147473541,
-	2147483647
-};
-
-/* Sine lookup table read */
-static inline int32_t sine_lookup(int idx)
+int32_t sin_fixed(int32_t th_rad_fxp)
 {
-	int32_t s;
-	int i1;
+	/*cordic_atan2_lookup_table = atan(2.^-(0:N-1)) N = 31 */
+	/*CORDIC Gain is cordic_gain = prod(sqrt(1 + 2.^(-2*(0:31-1)))) */
+	/*Inverse CORDIC Gain,inverse_cordic_gain = 1 / cordic_gain */
+	static const int32_t cordicsine_lookup[31] = { 843314857, 497837829, 263043837, 133525159,
+	 67021687, 33543516, 16775851, 8388437, 4194283, 2097149, 1048576, 524288,
+	 262144, 131072, 65536, 32768, 16384, 8192, 4096, 2048, 1024, 512, 256, 128,
+	 64, 32, 16, 8, 4, 2, 1 };
+	int32_t b_idx;
+	int32_t b_yn;
+	int32_t xn;
+	int32_t xtmp;
+	int32_t ytmp;
+	int32_t cdc_sin_th;
+	int sign;
+	/* 421657428 , deg = 90.000000 */
+	const int32_t CORDICSINE_PIOVERTWO_Q28FL  = Q_CONVERT_FLOAT(M_PI / 2, 28);
+	/* 843314857,  deg = 90.000000  */
+	const int32_t CORDICSINE_PIOVERTWO_Q29FL  = Q_CONVERT_FLOAT(M_PI / 2, 29);
+	/* 1686629713, deg = 90.000000  */
+	const int32_t CORDICSINE_PIOVERTWO_Q30FL  = Q_CONVERT_FLOAT(M_PI / 2, 30);
+	/* 652032874 , deg = 69.586061*/
+	const int32_t CORDICSINE_LUT_Q29FL  =   Q_CONVERT_FLOAT(1.214505869895220, 29);
+	/*Addition or subtraction by a multiple of pi/2 is done in the data type*/
+	/*of the input. When the fraction length is 29, then the quantization error*/
+	/*introduced by the addition or subtraction of pi/2 is done with 29 bits of*/
+	/*precision.Input range of cordicsin must be in the range [-2*pi, 2*pi),*/
+	/*a signed type with fractionLength = wordLength-4 will fit this range*/
+	/*without overflow.Increase of fractionLength makes the addition or */
+	/*subtraction of a multiple of pi/2 more precise */
 
-	i1 = idx & (2 * SINE_NQUART - 1);
-	if (i1 > SINE_NQUART)
-		i1 = 2 * SINE_NQUART - i1;
+	if (th_rad_fxp > CORDICSINE_PIOVERTWO_Q28FL) {
```

Comment:
Since the input range of cordicsin must be in the range [-2*pi, 2*pi), then we know that a signed type with fraction Length = wordLength-4 will fit this range without overflow.
In case you need additional param with a different input format, say Q8.24 ,then input will have less precision than Q4.28,
so upcasting to have the larger fraction Length does not change the value of the input, and acts like a left shift of the data. Thus increase of fraction Length makes the addition or subtraction of a multiple of pi/2 more precise.

Reply:
Im with @ShriramShastry, i think it is up to the caller to make sure they are handing the correct bit format

Reply:
@cujomalainey , now we have cordicsine output back to  Q1.31 format.